### PR TITLE
WM4_Dawit-Abraha_JavaScript_Core-2_Week-4

### PR DESCRIPTION
**Pull request by Dawit-Dev**
**Volunteers: Are you marking this coursework?** _You can find a guide on how to mark this coursework in `HOW_TO_MARK.md` in the root of this repository_

# Your Details

- Your Name:
- Your City:
- Your Slack Name:

# Homework Details

- Module:
- Week:

# Notes

- What did you find easy?

- What did you find hard?

- What do you still not understand?

- Any other notes?

### CODE REVIEW

File: 1-mandatory/1-alarmclock/alarmclock.js [hunk 1]
```diff
@@ -1,5 +1,26 @@
-function setAlarm() {}
+function setAlarm() {
+  let timeDisplay = document.getElementById("timeRemaining");
+  let inputTime = document.getElementById("alarmSet");
+  let seconds = inputTime.value;
```
Comment (jdbevan) [inline 1-mandatory/1-alarmclock/alarmclock.js:4]: What type is the `seconds` variable here? Object/string/number/Error/boolean?

File: 1-mandatory/1-alarmclock/alarmclock.js [hunk 2]
```diff
@@ -1,5 +1,26 @@
...
+  inputTime.value = "";
+  timeDisplay.innerText = `Time Remaining: ${timeFormatter(seconds)}`;
+  let countDownInterval = setInterval(() => {
+    seconds--;
+    timeDisplay.innerText = `Time Remaining: ${timeFormatter(seconds)}`;
+    if (timeDisplay.innerText === "Time Remaining: 00:00") {
```
Comment (jdbevan) [inline 1-mandatory/1-alarmclock/alarmclock.js:10]: What will you need to remember to do here if you change the output of `timeFormatter` to eg `${minutes}m ${seconds}s`? Can you think of a different way to achieve exactly the same conditional check?

File: 1-mandatory/1-alarmclock/alarmclock.js [hunk 3]
```diff
@@ -1,5 +1,26 @@
...
+      playAlarm();
+      clearInterval(countDownInterval);
+    }
+   }, 1000);
+}
 
+function timeFormatter(sec) {
```
Comment (jdbevan) [inline 1-mandatory/1-alarmclock/alarmclock.js:17]: Good use of a separate function 👍

File: 1-mandatory/2-quotegenerator/quotes.js [hunk 4]
```diff
@@ -1,5 +1,47 @@
...
+  let newQuote = pickFromArray(quotes);
+  quote.innerHTML = newQuote.quote;
+  author.innerHTML = newQuote.author;
+
+  btn.innerText = "New Quote";
+  btn.classList.remove("loading");
+});
+
+soundBtn.addEventListener("click", () => {
+  let utterance = new SpeechSynthesisUtterance(
```
Comment (jdbevan) [inline 1-mandatory/2-quotegenerator/quotes.js:23]: This is cool! I didn't know this existed :D

File: 1-mandatory/2-quotegenerator/quotes.js [hunk 5]
```diff
@@ -1,5 +1,47 @@
...
+
+copyBtn.addEventListener("click", () => {
+  navigator.clipboard.writeText(quote.innerHTML);
+});
+
+document
+  .querySelector(".switch input")
+  .addEventListener("change", (isChecked) => {
+    isChecked.target.checked
+      ? (interval = setInterval(() => {
```
Comment (jdbevan) [inline 1-mandatory/2-quotegenerator/quotes.js:37]: Where is `interval` first defined?

File: 1-mandatory/3-slideshow/slideshow.js [hunk 6]
```diff
@@ -1 +1,53 @@
...
+  "https://lh3.googleusercontent.com/A-26ppznGkhSnITdLtwoyyAuyLdYxl8gvBrQqX-V-yuR_gxc7-jJkyX1Sy7yrDob0k6b7MH8EadANw8gqzeANRN7ZoEwZImxw6Q6T6cVis-YErNP3Rj5FJLCybpvqx7GxsrukJnrUQBYs6Uuf9stYlqA2BUS0_zsKwy-UZKvCLhuXhVDQXeOpd6ndGeSwciAx-llyB8cgrkP0w3AWk1iEwpKGEH0UZGRVE9lLoUdegQpM-qbHpeTnFB7KqSqiwhh1AYzdQQYm82zSHx1qFoe7gy_X0p3C2jbrr4biWvS0P3L49iSLfZoPUfaTn4KVqHujFyIEQ2mQYYaQmdlUOL3_mRZJ6bELFYKy_gjR1LPU5qiFeznW91NnYru87fZ28Swql8eaTgFArv0ki7Dl4KnJcWMy_UczHkes5HieD_-oHfd-XZpCS_KdWILVCQnzloS1NOAsbzztnfwCClYFfT9b7xzNOo36UhhLHLeBx_YE1LOrBFZDXl4c6SnUnDxWc74U4dR4OYsgueJ3arEcmTpMk0OZiIboWskeXRpBYGQ6v8Pv3ojwwI9hJeUyL1V_ksCDHEE-XlILrSZ-uCr64F54oLtdOJ7QNDjeXb_DoDtrgnS9LS4isMSKEuYkLnYbV4lvr0DKxQmLtKJYt6YvSixQsEZ5YtR97RIY4wPPJRucVR3Ytbm8BIk48ZWb8buwWlercWI31XDjYQ15_fyOcnP31KI4ikeU3erqt_N8VHQ2bhX9QJs7CiALVsSej8u-A=w711-h948-no?authuser=0",
+];
+let element_image_1 = document.getElementById("image_1");
+element_image_1.setAttribute("src", images[0]);
+
+document.getElementById("next").addEventListener("click", () => {
+  images.push(images[0]);
```
Comment (jdbevan) [inline 1-mandatory/3-slideshow/slideshow.js:23]: This is quite a clever approach, but a little unconventional! I was expecting to see a variable which defined the array offset and was increased/decreased when the various buttons were clicked.
```
let imageNumber = 0;
function showImage() {
    element_image_1.setAttribute('src', images[imageNumber]);
}
function nextImage() {
    imageNumber = imageNumber === images.length ? 0 : imageNumber++;
}
function previousImage() {
    imageNumber = imageNumber === 0 ? images.length : imageNumber--;
}

```

File: 1-mandatory/3-slideshow/slideshow.js [hunk 7]
```diff
@@ -1 +1,53 @@
...
+  images.shift();
+  element_image_1.src = images[0];
```
Comment (jdbevan) [inline 1-mandatory/3-slideshow/slideshow.js:25]: Just curious why you set `src` directly here but you used `setAttribute` on line 20?

File: 1-mandatory/3-slideshow/slideshow.js [hunk 8]
```diff
@@ -1 +1,53 @@
...
+});
+
+document.getElementById("previous").addEventListener("click", () => {
+  images.unshift(images[images.length - 1]);
+  images.pop();
+  element_image_1.src = images[0];
+});
+
+let interval;
```
Comment (jdbevan) [inline 1-mandatory/3-slideshow/slideshow.js:34]: Great job defining the `interval` before setting/using it with the event listener 👍

**Review by jdbevan — COMMENTED** [review]
Great work! Be careful mixing/changing the type of a variable, and try keeping repeated logic in its own function

File: 1-mandatory/1-alarmclock/alarmclock.js (hunk 1 above)
Comment (Dawit-Dev) [reply 1-mandatory/1-alarmclock/alarmclock.js:4]: Number

File: 1-mandatory/2-quotegenerator/quotes.js (hunk 5 above)
Comment (Dawit-Dev) [reply 1-mandatory/2-quotegenerator/quotes.js:37]: Thank you for noticing for the error. I should have declared the interval globally before I applied inside the function.

File: 1-mandatory/3-slideshow/slideshow.js (hunk 7 above)
Comment (Dawit-Dev) [reply 1-mandatory/3-slideshow/slideshow.js:25]: Thank you, I should be consistent the way I assign attributes.

File: 1-mandatory/1-alarmclock/alarmclock.js (hunk 1 above)
Comment (jdbevan) [reply 1-mandatory/1-alarmclock/alarmclock.js:4]: Actually, the `value` of an `input` element is always a string as far as I can tell, regardless of the `type` attribute of the element. So you always need to pass it through `parseInt(input, 10)` and check for `NaN`

**Pull request closed without merging**